### PR TITLE
[Torch] NNCFNetwork.transformation_layout

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
On top of #2584 

### Changes

`NNCFNetwork.transformation_layout` method is introduced

### Reason for changes

To make it possible to collect applied transformation commands from a modified `NNCFNetwork`

### Related tickets

129586

### Tests

TestGetAppliedModificationCommands

### CODE REVIEW

File: tests/torch/helpers.py [hunk 1]
```diff
@@ -228,6 +351,66 @@ def num_flat_features(self, x):
...
+
+    def get_state(self):
+        return self._state
+
+    @classmethod
+    def from_state(cls, state: str):
+        return cls(state)
+
+
+def target_points_are_equal(tp_original: PTTargetPoint, tp_recovered: PTTargetPoint):
```
Comment (AlexanderDokuchaev) [inline tests/torch/helpers.py:375]: docstr

File: tests/torch/helpers.py [hunk 2]
```diff
@@ -228,6 +351,66 @@ def num_flat_features(self, x):
...
+    if tp_original != tp_recovered:
+        return False
+    if tp_original.target_type == TargetType.OPERATOR_PRE_HOOK:
+        return tp_original.input_port_id == tp_recovered.input_port_id
+    return True
+
+
+def are_commands_equal(
```
Comment (AlexanderDokuchaev) [inline tests/torch/helpers.py:383]: docstr

File: nncf/torch/nncf_network.py [hunk 3]
```diff
@@ -778,6 +784,119 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
...
+                        command_target_type = TargetType.OPERATION_WITH_WEIGHTS
+                        module = module.op
+                    if not isinstance(module, ExternalOpCallHook):
+                        command = _create_pt_insert_command(
+                            module, command_target_type, nncf_node.node_name, priority, None
+                        )
+                        transformation_layout.register(command)
+                        continue
+
+                    info = f"TargetType: {command_target_type}, nncf node name: {nncf_node.node_name},"
```
Comment (AlexanderDokuchaev) [inline nncf/torch/nncf_network.py:837]:
```
info = (
 "..."
 "..."
)
```

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: tests/torch/test_nncf_network.py [hunk 4]
```diff
@@ -1024,3 +1029,83 @@ def test_insert_at_point_hook_handles(self, target_type: TargetType, target_node
...
+
+
+@pytest.mark.parametrize("target_type", TwoConvTestModel.AVAILABLE_TARGET_TYPES)
+@pytest.mark.parametrize(
+    "command_builder,command_type", tuple(zip(TwoConvTestModel.get_command_builders(), TwoConvTestModel.COMMAND_TYPES))
+)
+def test_priority_of_get_applied_modification_commands(command_builder, target_type, command_type):
+    layout = PTTransformationLayout()
+    commands = dict()
+    for priority in (0, 3, 2, 4, 1):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:1066]: I need this priorities to test priorities are recovered correctly. As there are many possible ways to recover priorities (example: hooks == [op1, op2], then possible priorities could be (DEFAULT_PRIORITY, FP32_TENSOR_STATISTICS_OBSERVATION) or (FP32_TENSOR_STATISTICS_OBSERVATION, PRUNING_PRIORITY) or (SPARSIFICATION_PRIORITY, QUANTIZATION_PRIORITY) or (SPARSIFICATION_PRIORITY, SPARSIFICATION_PRIORITY)

I use index number as a priorities so the order of insertion is the same as my task is eventually recover model and is not to replicate exact commands which were used during compression

File: tests/torch/test_nncf_network.py [hunk 5]
```diff
@@ -1024,3 +1029,83 @@ def test_insert_at_point_hook_handles(self, target_type: TargetType, target_node
...
+def test_get_applied_modification_commands(command_builder, target_type):
+    command = command_builder(target_type, TransformationPriority.DEFAULT_PRIORITY)
+    if isinstance(command, PTSharedFnInsertionCommand) and target_type in [
+        TargetType.PRE_LAYER_OPERATION,
+        TargetType.POST_LAYER_OPERATION,
+    ]:
+        pytest.skip(f"PTSharedFnInsertionCommand is not supporting target type {target_type}")
+
+    model = TwoConvTestModel()
+    nncf_model = NNCFNetwork(deepcopy(model), input_info=FillerInputInfo([FillerInputElement([1, 1, 4, 4])]))
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:1045]: This won't affect the test in any way, I suggest to keep it as is. True and False parameters are checked in the model transformer tests, here the model transformer is used as a reference code

File: tests/torch/test_nncf_network.py [hunk 6]
```diff
@@ -1024,3 +1029,83 @@ def test_insert_at_point_hook_handles(self, target_type: TargetType, target_node
 
             del ref_hooks[-2]
             _check(ref_hooks)
+
+
+@pytest.mark.parametrize("target_type", TwoConvTestModel.AVAILABLE_TARGET_TYPES)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:1034]: What are the reasons and how to name the file? I'm testing a NNCFNetwork method, I believe `test_nncf_network` is a correct place for that

File: tests/torch/helpers.py [hunk 7]
```diff
@@ -198,6 +214,113 @@ def nz_weights_num(self):
     def nz_bias_num(self):
         return 2
 
+    @staticmethod
+    def create_pt_insertion_command(
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/helpers.py:218]: This methods are
1) Model specific (only works with TwoConvTestModel)
2) Are needed for other test in follow up pr (https://github.com/openvinotoolkit/nncf/pull/2531)

Please allow me to keep it as is

File: tests/torch/test_nncf_network.py [hunk 8]
```diff
@@ -1024,3 +1029,83 @@ def test_insert_at_point_hook_handles(self, target_type: TargetType, target_node
...
+    commands = dict()
+    for priority in (0, 3, 2, 4, 1):
+        if command_type is PTSharedFnInsertionCommand:
+            command = command_builder(target_type, priority, op_unique_name=f"UNIQUE_NAME_{priority}")
+        else:
+            command = command_builder(target_type, priority)
+        layout.register(command)
+        commands[priority] = command
+    else:
+        if isinstance(command, PTSharedFnInsertionCommand) and target_type in [
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py:1074]: What is the value of this refactoring?

File: nncf/torch/nncf_network.py [hunk 9]
```diff
@@ -778,6 +784,119 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
...
+        """
+        Collects all hooks applied to the NNCFNetwork, converts them to insertion commands
+        and returns in PTTransformationLayout format. Default hooks group name is used in
+        recovered commands, so hooks group names specified diring the model modification
+        become outdated.
+
+        :return: Transformation layout with all commands applied to the NNCFNetwork.
+        """
+
+        def _create_pt_insert_command(module, target_type, target_node_name, priority, input_port_id):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py:797]: Sure, I can copy/paste this code to two different places, but I preferer a nested function. Please suggest better solution if you have one

File: tests/torch/test_nncf_network.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py]: Done

File: tests/torch/test_nncf_network.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/test_nncf_network.py]: Done

File: tests/torch/helpers.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply tests/torch/helpers.py]: Done

File: nncf/torch/graph/graph.py [hunk 10]
```diff
@@ -60,6 +60,13 @@ def get_op_nodes_in_scope(self, scope: Scope) -> List[NNCFNode]:
                 matching_graph_op_nodes.extend(nodes_in_module)
         return matching_graph_op_nodes
 
+    def get_op_node_in_scope(self, scope: Scope) -> List[NNCFNode]:
+        for scope_str, nodes_in_module in self._layer_name_vs_shared_nodes.items():
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/graph/graph.py]: Done

File: nncf/torch/nncf_network.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Done

File: nncf/torch/nncf_network.py [hunk 11]
```diff
@@ -778,6 +784,119 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
...
+            target_point = PTTargetPoint(
+                target_type=target_type, target_node_name=target_node_name, input_port_id=input_port_id
+            )
+            return PTInsertionCommand(point=target_point, fn=module, priority=priority)
+
+        def _check_external_call_hook_is_valid(hook: ExternalOpCallHook, info: str):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py:797]: Done

File: nncf/torch/nncf_network.py [hunk 12]
```diff
@@ -778,6 +784,119 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
...
+            assert hasattr(
+                self, hook._storage_name
+            ), f"Storage name {hook._storage_name} is not registered. Info: {info}"
+            assert hook._storage_key in getattr(
+                self, hook._storage_name
+            ), f"Storage key {hook._storage_key} is not registered. Info: {info}"
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Done

File: tests/torch/helpers.py (hunk 1 above)
Comment (daniil-lyakhov) [reply tests/torch/helpers.py]: Done

File: tests/torch/helpers.py (hunk 2 above)
Comment (daniil-lyakhov) [reply tests/torch/helpers.py]: Done

File: nncf/torch/nncf_network.py (hunk 3 above)
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Done

File: nncf/torch/graph/graph.py [hunk 13]
```diff
@@ -60,6 +60,9 @@ def get_op_nodes_in_scope(self, scope: Scope) -> List[NNCFNode]:
                 matching_graph_op_nodes.extend(nodes_in_module)
         return matching_graph_op_nodes
 
+    def get_op_node_in_scope(self, scope: Scope) -> List[NNCFNode]:
+        return self._layer_name_vs_shared_nodes[str(scope)]
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/graph/graph.py:64]: `_layer_name_vs_shared_nodes` is a default dict with default list

File: tests/torch/nncf_network/helpers.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+            priority=priority,
+            hooks_group_name=group,
+        )
+
+    def get_command_builders(self):
+        """
+        Get all command builders available in a tuple.
+        """
+        return (
+            self.create_pt_insertion_command,
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/helpers.py:109]: return tuple (FN, Type of command) 
remove COMMAND_CLASSES

File: tests/torch/nncf_network/helpers.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+                compression_module_type=ExtraCompressionModuleType.EXTERNAL_OP,
+            ),
+            functools.partial(
+                self.create_pt_shared_fn_insertion_command,
+                compression_module_type=ExtraCompressionModuleType.EXTERNAL_QUANTIZER,
+            ),
+        )
+
+    COMMAND_CLASSES = [PTInsertionCommand, PTSharedFnInsertionCommand, PTSharedFnInsertionCommand]
+    PRIORITIES = (TransformationPriority.QUANTIZATION_PRIORITY, TransformationPriority.QUANTIZATION_PRIORITY.value + 1)
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/helpers.py:121]: need comment why  used .value + 1

File: tests/torch/nncf_network/test_get_applied_modifications.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+
+@pytest.mark.parametrize(
+    "trace_parameters_p,model_cls,target_type,command_builder",
+    _get_trace_params_target_types_command_builders_and_models_cls(),
+)
+def test_priority_of_get_applied_modification_commands(command_builder, model_cls, target_type, trace_parameters_p):
+    layout = PTTransformationLayout()
+    commands = dict()
+    for priority in (0, 3, 2, 4, 1):
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/test_get_applied_modifications.py:123]: is not enough 3?

File: tests/torch/nncf_network/test_get_applied_modifications.py [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+@pytest.mark.parametrize("model_cls", MODELS_TO_TEST)
+def test_all_possible_combinations_of_commands_for_get_applied_commands(model_cls, trace_parameters):
+    dummy_state = "DummyState"
+    commands = InsertionCommandBuilder(model_cls).get_all_available_commands(
+        dummy_state, skip_model_transformer_unsupported=not trace_parameters, trace_parameters=trace_parameters
+    )
+
+    model = model_cls()
+    nncf_model = wrap_model(model, torch.zeros([1, 1, 4, 4]), trace_parameters=trace_parameters)
+    model_tranformer = PTModelTransformer(nncf_model)
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/test_get_applied_modifications.py:153]:
```suggestion
    model_transformer = PTModelTransformer(nncf_model)
```

File: tests/torch/nncf_network/test_get_applied_modifications.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+        if sum(map(int, eq_commands)) != 1:
+            raise RuntimeError(f"Command {command} has no pair in recovered commands")
+
+
+@pytest.mark.parametrize("target_type", (TargetType.OPERATION_WITH_WEIGHTS, TargetType.OPERATOR_PRE_HOOK))
+@pytest.mark.parametrize("model_cls", MODELS_TO_TEST)
+def test_get_applied_modification_commands_broken_call_hook(model_cls, target_type, trace_parameters):
+    model = model_cls()
+    nncf_model = wrap_model(model, torch.zeros([1, 1, 4, 4]), trace_parameters=trace_parameters)
+    model_tranformer = PTModelTransformer(nncf_model)
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/test_get_applied_modifications.py:174]:
```suggestion
    model_transformer = PTModelTransformer(nncf_model)
```

File: tests/torch/nncf_network/test_get_applied_modifications.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+@pytest.fixture(name="trace_parameters", params=(True, False))
+def trace_parameters_fixture(request) -> bool:
+    return request.param
+
+
+MODELS_TO_TEST = (TwoConvTestModel, TwoSharedConvTestModel)
+
+
+def _get_trace_params_target_types_command_builders_and_models_cls():
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/test_get_applied_modifications.py:48]: docstring and typehint

File: tests/torch/nncf_network/test_get_applied_modifications.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+        target_point.target_type = new_target_type
+
+
+@pytest.mark.parametrize(
+    "trace_parameters_p,model_cls,target_type,command_builder",
+    _get_trace_params_target_types_command_builders_and_models_cls(),
+)
+def test_get_applied_modification_commands(model_cls, command_builder, target_type, trace_parameters_p):
+    model = model_cls()
+    nncf_model = wrap_model(model, torch.zeros([1, 1, 4, 4]), trace_parameters=trace_parameters_p)
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/test_get_applied_modifications.py:100]:
```suggestion
    nncf_model = wrap_model(model, torch.zeros(model_cls.INPUT_SHAPE), trace_parameters=trace_parameters_p)
```

File: nncf/torch/nncf_network.py [hunk 21]
```diff
@@ -778,6 +784,127 @@ def _collect_eval_op_scopes(self, model: nn.Module, dummy_forward_fn: Callable)
                 result.append(scope_in_model)
         return result
 
+    def get_applied_transformation_layout(self) -> PTTransformationLayout:
+        """
+        Collects all hooks applied to the NNCFNetwork, converts them to insertion commands
+        and returns in PTTransformationLayout format. Default hooks group name is used in
+        recovered commands, so hooks group names specified diring the model modification
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Done, thanks

File: tests/torch/nncf_network/helpers.py (hunk 14 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/helpers.py]: Done

File: tests/torch/nncf_network/helpers.py (hunk 15 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/helpers.py:133]: Done

File: tests/torch/nncf_network/test_get_applied_modifications.py (hunk 16 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/test_get_applied_modifications.py]: Good catch, done

File: tests/torch/nncf_network/test_get_applied_modifications.py (hunk 17 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/test_get_applied_modifications.py]: Done

File: tests/torch/nncf_network/test_get_applied_modifications.py (hunk 18 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/test_get_applied_modifications.py]: Done

File: tests/torch/nncf_network/test_get_applied_modifications.py (hunk 19 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/test_get_applied_modifications.py]: Done

File: tests/torch/nncf_network/test_get_applied_modifications.py (hunk 20 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/test_get_applied_modifications.py]: Done

File: tests/torch/nncf_network/helpers.py [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+            (
+                functools.partial(
+                    self.create_pt_shared_fn_insertion_command,
+                    compression_module_type=ExtraCompressionModuleType.EXTERNAL_QUANTIZER,
+                ),
+                PTSharedFnInsertionCommand,
+            ),
+        )
+
+    COMMAND_CLASSES = [PTInsertionCommand, PTSharedFnInsertionCommand, PTSharedFnInsertionCommand]
```
Comment (AlexanderDokuchaev) [inline tests/torch/nncf_network/helpers.py:147]: unused?

**Review by AlexanderDokuchaev — APPROVED** [review]

File: tests/torch/nncf_network/helpers.py (hunk 22 above)
Comment (daniil-lyakhov) [reply tests/torch/nncf_network/helpers.py]: Yes, I've removed it

File: nncf/torch/nncf_network.py [hunk 23]
```diff
@@ -1132,3 +1270,26 @@ def compression_module_type_to_attr_name(compression_module_type: ExtraCompressi
     if compression_module_type == ExtraCompressionModuleType.EXTERNAL_OP:
         return EXTERNAL_OP_STORAGE_NAME
     raise nncf.ValidationError("Unknown extra module type")
+
+
+def create_pt_insertion_command(
```
Comment (alexsu52) [inline nncf/torch/nncf_network.py:1275]: What do you think about moving this function to `commands.py`?

**Review by alexsu52 — COMMENTED** [review]
LGTM

File: nncf/torch/nncf_network.py (hunk 23 above)
Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Agree, done

Comment (daniil-lyakhov) [reply nncf/torch/nncf_network.py]: Moved to `command_creation.py`

**Pull request merged**